clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -849,9 +953,22 @@ public OptionalLong currentLag(TopicPartition topicPartition) {
         }
     }
 
+    public void setGroupMetadata(final ConsumerGroupMetadata groupMetadata) {
+        this.groupMetadata = Optional.of(groupMetadata);
+    }
+
     @Override
     public ConsumerGroupMetadata groupMetadata() {
-        throw new KafkaException("method not implemented");
+        acquireAndEnsureOpen();
+        try {
+            maybeThrowInvalidGroupIdException();
+            backgroundEventProcessor.process();
+            return groupMetadata.orElseThrow(
+                () -> new IllegalStateException("No group metadata found although a valid group ID exists. This is a bug!")
```

Comment:
I'm not sure what we should throw here. Does the API contract allow for `IllegalStateException`?

Reply:
You *know* that groupMetadata is present because of the earlier `maybeThrowInvalidGroupIdException`. I suppose one pattern would be to return an unwrapped `GroupMetadata` from maybeThrowInvalidGroupIdException so that you've eliminated the possibility of the `Optional` being empty.

Reply:
Actually, `backgroundEventProcessor.process();` sets the group metadata. So returning group metadata from `maybeThrowInvalidGroupIdException` might return an outdated group metadata. Since now I removed `backgroundEventProcessor.process();`. I will reconsider the exception.